## PCA continued ..

In [14]:
from sklearn.datasets import fetch_openml
from sklearn.decomposition import PCA

mnist = fetch_openml('mnist_784', as_frame = False)
X_train, y_train = mnist.data[:60_000], mnist.target[:60_000]
X_test, y_test = mnist.data[60_000:], mnist.target[60_000:]

pca = PCA(n_components = 0.95)
X_reduced = pca.fit_transform(X_train)
pca.n_components_

/opt/anaconda3/lib/python3.11/site-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


154

This states we need **154** features to preserve variance by 0.95

### Finding Best no. of components with Radomized Search

In [25]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import RandomizedSearchCV
from sklearn.pipeline import make_pipeline
import numpy as np

clf = make_pipeline(PCA(random_state=42), RandomForestClassifier(random_state=42))
param_dist = {
    'pca__n_components' : np.arange(10,80),
    'randomforestclassifier__n_estimators' : np.arange(50, 500)
}
rand_search = RandomizedSearchCV(clf, param_dist, n_iter=10, cv=3, random_state=42)
rand_search.fit(X_train[:1000], y_train[:1000])
rand_search.best_params_

{'randomforestclassifier__n_estimators': 465, 'pca__n_components': 23}

So, the 784 dimensional data is reduced to 23 components using PCA. 